week_2/pet.rb

```diff
@@ -1,115 +1,132 @@
+    @name = name
+    @animal_type_id = animal_type_id
+    @food_consumed_per_day = food_consumed_per_day
+  end
+
+  # Return the habitat of the pet
+  def habitat
+    HABITATS.values.each do |x|
+      x.each do |k|
+        if(k==animal_type_id)
+          return HABITATS.key(x)
+        end
+      end
+    end
+  end
+
+  # Returns the cost of food required to feed the animal 
+  # per day
+  def food_cost_per_day
+    return FOOD_COST_PER_KG[animal_type_id] * food_consumed_per_day
+  end
+
+  # This function takes the number of `days` as the input
+  # and returns the food required(in kgs) to feed the
+  # pet for that many number of days
+  #
+  # Eg - if the pet requires 0.5 kg of food per day
+  # (i.e, food_consumed_per_day = 0.5), then
+  # pet.food_required(5) will return 2.5 as the answer
+  #
+  # cat = Pet.new(name: 'cat', animal_type_id: 6, food_consumed_per_day: 0.4)
+  # cat.food_required(28) = 11.2 (0.4 * 28)
+  def food_required(days)
+    return food_consumed_per_day * days
+  end
+
+  # This function takes the number of `days` as the input
+  # and returns the cost incurred to feed the pet for
+  # that many number of days
+  #
+  # cat = Pet.new(name: 'cat', animal_type_id: 6, food_consumed_per_day: 0.4)
+  # cat.food_cost(28) = 8960
+  def food_cost(days)
+    return FOOD_COST_PER_KG[animal_type_id] * food_consumed_per_day * days
+  end
+
+  # This function takes an array of pets and the `days`
+  # as input and returns the cost to feed them all 
+  # for the specified number of `days`
+  #
+  # Eg -
+  # cat = Pet.new(name: 'cat', animal_type_id: 6, food_consumed_per_day: 0.4)
+  # dog = Pet.new(name: 'dog', animal_type_id: 6, food_consumed_per_day: 0.7)
+  # fish = Pet.new(name: 'clownfish', animal_type_id: 2, food_consumed_per_day: 0.1)
+  # snake = Pet.new(name: 'python', animal_type_id: 4, food_consumed_per_day: 0.3)
+  # Pet.cost_to_feed([cat, dog, fish, snake], 6) will return 6180.0
+  def self.cost_to_feed(pets, days)
+    sum = 0
+    pets.each{|x|
+      sum += FOOD_COST_PER_KG[x.animal_type_id]*x.food_consumed_per_day*days
+    }
+    return sum
+  end
+
+  # This function takes an array of pets as input
+  # and returns a hash with pets name grouped by their animal type
+  #
+  # Eg - 
+  # cat = Pet.new(name: 'cat', animal_type_id: 6, food_consumed_per_day: 0.4)
+  # dog = Pet.new(name: 'dog', animal_type_id: 6, food_consumed_per_day: 0.7)
+  # fish = Pet.new(name: 'clownfish', animal_type_id: 2, food_consumed_per_day: 0.1)
+  # snake = Pet.new(name: 'python', animal_type_id: 4, food_consumed_per_day: 0.3)
+  # Pet.group_by_animal_type([cat, dog, fish, snake]) will return the follwing hash
+  # {
+  #   6 => ['cat', 'dog'],
+  #   2 => ['clownfish'],
+  #   4 => ['python']
+  # }
+  #
+  # Note - Order is not important
+  def self.group_by_animal_type(pets)
+    names = {}
+    pets.each{|x|
+      names[x.animal_type_id] = [];
+    }
+    pets.each{|x|
+      names[x.animal_type_id].push(x.name)
+    }

```

Comment:
This can done in a single loop by checking if a key exists in the has.
```ruby
if names[x.animal_type_id]
  # ....
else
  names[x.animal_type_id] = [x.name]
end
```

Reply:
ahh yes 👍